code/WorkInProgress/DrMelonsStuff.dm

```diff
@@ -114,117 +147,178 @@
+				on = TRUE
+
+	proc/drain_bathtub(mob/user as mob)
+		src.add_fingerprint(user)
+		if (get_dist(usr, src) <= 1 && !is_incapacitated(usr))
+			if (src.reagents.total_volume)
+				user.visible_message("<span class='notice'>[user] reaches into the bath and pulls the plug.", "<span class='notice'>You reach into the bath and pull the plug.</span>")
+				if (ishuman(usr))
+					var/mob/living/carbon/human/H = usr
+					if(!H.gloves)
+						reagents.reaction(H, TOUCH)
+				playsound(src.loc, "sound/misc/drain_glug.ogg", 70, 1)
+				src.reagents.clear_reagents()
+				src.on_reagent_change()
+
+				var/count = 0
+				for (var/obj/O in src)
+					count++
+					qdel(O)
+				if (count > 0)
+					user.visible_message("<span class='alert'>...and something flushes down the drain. Damn!", "<span class='alert'>...and flush something down the drain. Damn!</span>")
+			else
+				boutput(usr, "<span class='notice'>The bathtub's already empty.</span>")
 
 	relaymove(mob/user as mob, dir)
-		src.eject_user()
-		boutput(user, "<span class='notice'>You get out of the bath.</span>")
+		src.eject_occupant(user)
+
+	mouse_drop(atom/over_object as obj)
+		if (istype(over_object, /turf))
+			drain_bathtub(usr)
+			return
+		if (!istype(over_object, /obj/item/reagent_containers/glass) && !istype(over_object, /obj/item/reagent_containers/food/drinks) && !istype(over_object, /obj/reagent_dispensers) && !istype(over_object, /obj/item/spraybottle) && !istype(over_object, /obj/machinery/plantpot) && !istype(over_object, /obj/mopbucket) && !istype(over_object, /obj/item/reagent_containers/mender) && !istype(over_object, /obj/item/tank/jetpack/backtank) && !istype(over_object, /obj/machinery/bathtub))
+			return ..()
+		if (usr.stat || usr.getStatusDuration("weakened") || BOUNDS_DIST(usr, src) > 0 || BOUNDS_DIST(usr, over_object) > 0)  //why has this bug been in since i joined goonstation and nobody even looked here yet wtf -ZeWaka
+			boutput(usr, "<span class='alert'>That's too far!</span>")
+			return
 
+		src.transfer_all_reagents(over_object, usr)
 
 	process()
-		if (src.myuser)
-			if(src.myuser.loc != src)
-				src.myuser.pixel_y = 0
-				src.myuser = null
-				src.overlays = null
-				on_reagent_change()
-
+		if (src.on)
+			src.reagents.add_reagent(src.default_reagent, 100)
+			src.on_reagent_change()
+			if (src.reagents.is_full())
+				src.visible_message("<span class='notice'>As the [src] finishes filling, the tap shuts off automatically.</span>")
+				playsound(src.loc, "sound/misc/pourdrink2.ogg", 60, 5)
+				src.on = FALSE
+		if (src.occupant)
+			if(src.occupant.loc != src)
+				src.occupant.pixel_y = 0
+				src.occupant = null
+				src.on_reagent_change()
 				return
-			if(src.reagents.total_volume > 5)
-				reagents.reaction(src.myuser, TOUCH)
-				src.reagents.trans_to(src.myuser,5,1)
-			else
-				src.reagents.clear_reagents()
-				on_reagent_change()
-
-		return
+			if(src.reagents.total_volume)
+				src.reagents.reaction(src.occupant, TOUCH)
```

Comment:
Again, should probably have a reaction volume, and not just apply all 500u. Also should use the returned list from `reaction()` to consume reagents with an on-touch effect like styptic/sulf/acid/etc

Reply:
Reworked this to match how fluids handle touch reactions to match behavior